File: src/models/qmdd.m
#import "QMDD.h"
#import "UniqueTable.h"
#import "Calculation.h"

@implementation QMDDEdge

- (instancetype)initWithWeight:(simd_double2)weight node:(QMDDNode *)node {
    self = [super init];
    if (self) {
        _weight = weight;
        _uniqueTableKey = node ? [Calculation generateUniqueTableKey:node] : 0;
        _isTerminal = !node;
        UniqueTable *table = [UniqueTable sharedInstance];
        if (![table find:_uniqueTableKey] && node) {
            [table insert:_uniqueTableKey node:node];
        }
    }
    return self;
}

- (instancetype)initWithWeight:(double)weight node:(QMDDNode *)node {
    return [self initWithWeight:simd_make_double2(weight, 0.0) node:node];
}

- (instancetype)initWithWeight:(simd_double2)weight key:(NSUInteger)key {
    self = [super init];
    if (self) {
        _weight = weight;
        _uniqueTableKey = key;
        _isTerminal = NO;
    }
    return self;
}

- (instancetype)initWithWeight:(double)weight key:(NSUInteger)key {
    return [self initWithWeight:simd_make_double2(weight, 0.0) key:key];
}

- (QMDDNode *)getStartNode {
    UniqueTable *table = [UniqueTable sharedInstance];
    return [table find:_uniqueTableKey];
}

- (NSArray<NSValue *> *)getAllElementsForKet {
    UniqueTable *table = [UniqueTable sharedInstance];
    QMDDNode *node = [table find:_uniqueTableKey];
    if (node.edges.count == 1) {
        @throw [NSException exceptionWithName:@"RuntimeException" reason:@"The start node has only one edge, which is not allowed." userInfo:nil];
    } else {
        NSMutableArray<NSValue *> *result = [NSMutableArray array];
        if (_isTerminal) {
            for (NSArray<QMDDEdge *> *edgeRow in node.edges) {
                for (QMDDEdge *edge in edgeRow) {
                    [result addObject:[NSValue valueWithBytes:&(simd_mul(_weight, edge.weight)) objCType:@encode(simd_double2)]];
                }
            }
        } else {
            NSArray<NSValue *> *child0 = [node.edges[0][0] getAllElementsForKet];
            NSArray<NSValue *> *child1 = [node.edges[1][0] getAllElementsForKet];
            [result addObjectsFromArray:child0];
            [result addObjectsFromArray:child1];
            for (NSUInteger i = 0; i < result.count; i++) {
                simd_double2 value;
                [result[i] getValue:&value];
                value = simd_mul(_weight, value);
                result[i] = [NSValue valueWithBytes:&value objCType:@encode(simd_double2)];
            }
        }
        return result;
    }
}

- (BOOL)isEqualToEdge:(QMDDEdge *)other {
    if (!other) {
        return NO;
    }
    if (!simd_equal(_weight, other.weight)) {
        return NO;
    }
    if (_isTerminal != other.isTerminal) {
        return NO;
    }
    if (!_isTerminal && _uniqueTableKey != other.uniqueTableKey) {
        return NO;
    }
    UniqueTable *table = [UniqueTable sharedInstance];
    if (!_isTerminal && [table find:_uniqueTableKey] != [table find:other.uniqueTableKey]) {
        return NO;
    }
    return YES;
}

@end

@implementation QMDDNode

- (instancetype)initWithEdges:(NSArray<NSArray<QMDDEdge *> *> *)edges {
    self = [super init];
    if (self) {
        _edges = edges;
    }
    return self;
}

- (BOOL)isEqualToNode:(QMDDNode *)other {
    if (!other) {
        return NO;
    }
    if (_edges.count != other.edges.count) {
        return NO;
    }
    for (NSUInteger i = 0; i < _edges.count; i++) {
        if (_edges[i].count != other.edges[i].count) {
            return NO;
        }
        for (NSUInteger j = 0; j < _edges[i].count; j++) {
            if (![_edges[i][j] isEqualToEdge:other.edges[i][j]]) {
                return NO;
            }
        }
    }
    return YES;
}

@end

@implementation QMDDGate

- (instancetype)initWithEdge:(QMDDEdge *)edge numEdge:(NSUInteger)numEdge {
    self = [super init];
    if (self) {
        _initialEdge = edge;
        _depth = 0;
        [self calculateDepth];
    }
    return self;
}

- (void)calculateDepth {
    UniqueTable *table = [UniqueTable sharedInstance];
    QMDDNode *currentNode = [table find:_initialEdge.uniqueTableKey];
    NSUInteger currentDepth = 0;

    while (currentNode && currentNode.edges.count > 0) {
        currentDepth++;
        currentNode = [table find:currentNode.edges[0][0].uniqueTableKey];
    }
    _depth = currentDepth;
}

- (QMDDNode *)getStartNode {
    UniqueTable *table = [UniqueTable sharedInstance];
    return [table find:_initialEdge.uniqueTableKey];
}

- (QMDDEdge *)getInitialEdge {
    return _initialEdge;
}

- (NSUInteger)getDepth {
    return _depth;
}

- (BOOL)isEqualToGate:(QMDDGate *)other {
    if (!other) {
        return NO;
    }
    if (![_initialEdge isEqualToEdge:other.initialEdge]) {
        return NO;
    }
    if (_depth != other.depth) {
        return NO;
    }
    return YES;
}

@end

@implementation QMDDState

- (instancetype)initWithEdge:(QMDDEdge *)edge {
    self = [super init];
    if (self) {
        _initialEdge = edge;
        _depth = 0;
        [self calculateDepth];
    }
    return self;
}

- (void)calculateDepth {
    UniqueTable *table = [UniqueTable sharedInstance];
    QMDDNode *currentNode = [table find:_initialEdge.uniqueTableKey];
    NSUInteger currentDepth = 0;

    while (currentNode && currentNode.edges.count > 0) {
        currentDepth++;
        currentNode = [table find:currentNode.edges[0][0].uniqueTableKey];
    }
    _depth = currentDepth;
}

- (QMDDNode *)getStartNode {
    UniqueTable *table = [UniqueTable sharedInstance];
    return [table find:_initialEdge.uniqueTableKey];
}

- (QMDDEdge *)getInitialEdge {
    return _initialEdge;
}

- (NSUInteger)getDepth {
    return _depth;
}

- (NSArray<NSValue *> *)getAllElements {
    return [_initialEdge getAllElementsForKet];
}

- (QMDDState *)addState:(QMDDState *)other {
    QMDDNode *newNode = [self addNodes:[self getStartNode] node2:[other getStartNode]];
    return [[QMDDState alloc] initWithEdge:[[QMDDEdge alloc] initWithWeight:simd_add(_initialEdge.weight, other.initialEdge.weight) node:newNode]];
}

- (QMDDNode *)addNodes:(QMDDNode *)node1 node2:(QMDDNode *)node2 {
    UniqueTable *table = [UniqueTable sharedInstance];
    if (!node1) {
        return node2;
    }
    if (!node2) {
        return node1;
    }

    NSMutableArray<NSMutableArray<QMDDEdge *> *> *resultEdges = [NSMutableArray arrayWithCapacity:node1.edges.count];
    for (NSUInteger i = 0; i < node1.edges.count; i++) {
        NSMutableArray<QMDDEdge *> *row = [NSMutableArray arrayWithCapacity:node1.edges[i].count];
        for (NSUInteger j = 0; j < node1.edges[i].count; j++) {
            QMDDEdge *edge1 = node1.edges[i][j];
            QMDDEdge *edge2 = node2.edges[i][j];
            QMDDEdge *newEdge = [[QMDDEdge alloc] initWithWeight:simd_add(edge1.weight, edge2.weight) node:[self addNodes:[table find:edge1.uniqueTableKey] node2:[table find:edge2.uniqueTableKey]]];
            [row addObject:newEdge];
        }
        [resultEdges addObject:row];
    }

    return [[QMDDNode alloc] initWithEdges:resultEdges];
}

- (BOOL)isEqualToState:(QMDDState *)other {
    if (!other) {
        return NO;
    }
    if (![_initialEdge isEqualToEdge:other.initialEdge]) {
        return NO;
    }
    return YES;
}

@end

File: src/models/qmdd.go
package main

import (
    "fmt"
    "math/cmplx"
)

type QMDDNode struct {
    Edges [][]QMDDEdge
}

type QMDDEdge struct {
    Weight        complex128
    UniqueTableKey uint64
    IsTerminal    bool
}

func NewQMDDEdge(weight complex128, node *QMDDNode) QMDDEdge {
    var key uint64
    if node != nil {
        key = generateUniqueTableKey(*node)
    }
    return QMDDEdge{
        Weight:        weight,
        UniqueTableKey: key,
        IsTerminal:    node == nil,
    }
}

func NewQMDDEdgeWithComplex(w complex128, n *QMDDNode) *QMDDEdge {
    uniqueTableKey := uint64(0)
    isTerminal := n == nil
    if !isTerminal {
        uniqueTableKey = generateUniqueTableKey(n)
        table := GetUniqueTableInstance()
        if table.Find(uniqueTableKey) == nil {
            table.Insert(uniqueTableKey, n)
        }
    }
    return &QMDDEdge{weight: w, uniqueTableKey: uniqueTableKey, isTerminal: isTerminal}
}

func NewQMDDEdgeWithDouble(w float64, n *QMDDNode) *QMDDEdge {
    return NewQMDDEdgeWithComplex(complex(w, 0), n)
}

func NewQMDDEdgeWithComplexAndKey(w complex128, key uint64) *QMDDEdge {
    isTerminal := key == 0
    return &QMDDEdge{weight: w, uniqueTableKey: key, isTerminal: isTerminal}
}

func NewQMDDEdgeWithDoubleAndKey(w float64, key uint64) *QMDDEdge {
    return NewQMDDEdgeWithComplexAndKey(complex(w, 0), key)
}

func generateUniqueTableKey(node QMDDNode) uint64 {
    // ユニークなキーを生成するロジックをここに実装
    return 0 // 仮の値
}

func (edge QMDDEdge) GetStartNode() *QMDDNode {
    // UniqueTableからノードを取得するロジックをここに実装
    return nil // 仮の値
}

func (edge QMDDEdge) GetAllElementsForKet() []complex128 {
    var result []complex128
    var nodeStack []struct {
        Node      *QMDDNode
        EdgeIndex int
    }

    if edge.IsTerminal {
        result = append(result, edge.Weight)
    } else {
        nodeStack = append(nodeStack, struct {
            Node      *QMDDNode
            EdgeIndex int
        }{edge.GetStartNode(), 0})

        for len(nodeStack) > 0 {
            top := nodeStack[len(nodeStack)-1]
            nodeStack = nodeStack[:len(nodeStack)-1]

            if len(top.Node.Edges) == 1 {
                panic("The start node has only one edge, which is not allowed.")
            }

            for i := top.EdgeIndex; i < len(top.Node.Edges); i++ {
                if top.Node.Edges[i][0].IsTerminal {
                    result = append(result, top.Node.Edges[i][0].Weight)
                } else {
                    nodeStack = append(nodeStack, struct {
                        Node      *QMDDNode
                        EdgeIndex int
                    }{top.Node, i + 1})
                    nodeStack = append(nodeStack, struct {
                        Node      *QMDDNode
                        EdgeIndex int
                    }{top.Node.Edges[i][0].GetStartNode(), 0})
                    break
                }
            }
        }
    }
    return result
}

func (edge QMDDEdge) Equals(other QMDDEdge) bool {
    if edge.Weight != other.Weight {
        return false
    }
    if edge.IsTerminal != other.IsTerminal {
        return false
    }
    if !edge.IsTerminal && edge.UniqueTableKey != other.UniqueTableKey {
        return false
    }
    // UniqueTableからノードを比較するロジックをここに実装
    return true
}

func (edge QMDDEdge) NotEquals(other QMDDEdge) bool {
    return !edge.Equals(other)
}

func (edge QMDDEdge) String() string {
    return fmt.Sprintf("Weight = %v, Key = %v, isTerminal = %v", edge.Weight, edge.UniqueTableKey, edge.IsTerminal)
}

func NewQMDDNode(edges [][]QMDDEdge) *QMDDNode {
    return &QMDDNode{Edges: edges}
}

func (node *QMDDNode) Equals(other *QMDDNode) bool {
    if len(node.Edges) != len(other.Edges) {
        return false
    }
    for i := range node.Edges {
        if len(node.Edges[i]) != len(other.Edges[i]) {
            return false
        }
        for j := range node.Edges[i] {
            if !node.Edges[i][j].Equals(other.Edges[i][j]) {
                return false
            }
        }
    }
    return true
}

func (node *QMDDNode) NotEquals(other *QMDDNode) bool {
    return !node.Equals(other)
}

func (node *QMDDNode) String() string {
    result := fmt.Sprintf("Node with %d rows of edges\n", len(node.Edges))
    for i, row := range node.Edges {
        for j, edge := range row {
            result += fmt.Sprintf("    Edge (%d, %d): %s\n", i, j, edge.String())
        }
    }
    return result
}

type QMDDGate struct {
    InitialEdge QMDDEdge
}

func NewQMDDGate(edge QMDDEdge) *QMDDGate {
    return &QMDDGate{InitialEdge: edge}
}

func (gate *QMDDGate) GetStartNode() *QMDDNode {
    return gate.InitialEdge.GetStartNode()
}

func (gate *QMDDGate) GetInitialEdge() QMDDEdge {
    return gate.InitialEdge
}

func (gate *QMDDGate) Equals(other *QMDDGate) bool {
    return gate.InitialEdge.Equals(other.InitialEdge)
}

func (gate *QMDDGate) NotEquals(other *QMDDGate) bool {
    return !gate.Equals(other)
}

func (gate *QMDDGate) String() string {
    return fmt.Sprintf("Gate with initial edge: %s", gate.InitialEdge.String())
}

type QMDDState struct {
    InitialEdge QMDDEdge
}

func NewQMDDState(edge QMDDEdge) *QMDDState {
    return &QMDDState{InitialEdge: edge}
}

func (state *QMDDState) GetStartNode() *QMDDNode {
    return state.InitialEdge.GetStartNode()
}

func (state *QMDDState) GetInitialEdge() QMDDEdge {
    return state.InitialEdge
}

func (state *QMDDState) GetAllElements() []complex128 {
    return state.InitialEdge.GetAllElementsForKet()
}

func (state *QMDDState) Equals(other *QMDDState) bool {
    return state.InitialEdge.Equals(other.InitialEdge)
}

func (state *QMDDState) NotEquals(other *QMDDState) bool {
    return !state.Equals(other)
}

func (state *QMDDState) String() string {
    return fmt.Sprintf("State with initial edge: %s", state.InitialEdge.String())
}

